problem/create-your-own-Promise_en.md

```diff
@@ -1,4 +1,1090 @@
+  // to do
+
+  return new MyPromise((resolve, reject) => {});
+}
+```
+
+1. **Handling the first argument in our `then()` method**
+
+   Since the callback function runs when the promise is resolved, it cannot be executed within the `then()` method.
+
+   For example:
+
+   ```js
+   class MyPromise {
+     #state;
+     #result;
+
+     constructor(executor) {
+       this.#state = 'pending';
+       this.#result = undefined;
+
+       try {
+         executor(this.#resolve.bind(this), this.#reject.bind(this));
+       } catch (error) {
+         this.#reject(error);
+       }
+     }
+
+     #resolve(value) {
+       // ...
+       this.#state = 'fulfilled';
+       this.#result = value;
+     }
+
+     then(onFulfilled) {
+       onFulfilled(this.#result);
+     }
+   }
+
+   const p = new MyPromise((resolve) => {
+     resolve(10);
+   }).then((value) => {
+     console.log(value); // 10
+   });
+   ```
+
+   Although the code above seems to work, it doesn't work as intended when the promise is resolved asynchronously, even if the `onFulfilled()` is called asynchronously:
+
+   ```js
+   class MyPromise {
+     // ...
+
+     then(onFulfilled) {
+       // Call `onFulfilled()` asynchronously.
+       queueMicrotask(() => {
+         onFulfilled(this.#result);
+       });
+     }
+   }
+
+   const p = new MyPromise((resolve) => {
+     setTimeout(() => {
+       resolve(10);
+     }, 0);
+   }).then((value) => {
+     console.log(value); // undefined
+   });
+   ```
+
+   Therefore, the `onFulfilled()` function should be called within the `#resolve()` method and the `then()` method just registers the `onFulfilled()` function. The `onFulfilled()` function is like a subscriber, subscribing to the promised result, and the `then()` method is kind of like the function `subscribe()` in the Publisher/Subscriber Pattern, which receives subscriber callbacks and stores/registers them in some data structures.
+
+   ```js
+   class MyPromise {
+     // ...
+     #onFulfilled;
+
+     constructor(executor) {
+       // ...
+       this.#onFulfilled = undefined;
+       // ...
+     }
+
+     #resolve(value) {
+       //...
+       this.#onFulfilled(this.#result);
+     }
+
+     // ...
+
+     then(onFulfilled) {
+       // If `onFulfilled` is not a function, replace it with a function
+       // that simply returns the received result.
+       this.#onFulfilled =
+         typeof onFulfilled === 'function' ? onFulfilled : (value) => value;
+
+       return new Promise((resolve, reject) => {});
+     }
+   }
+   ```
+
+   Although the `then` method will be triggered instantly, the callback functions (handlers) will be invoked asynchronously. `Promise` uses the _microtask_ queue to run the callbacks. When a promise is settled, its `then` handlers are add into the microtask queue. [The microtasks get run whenever JavaScript finishes executing, in other words, whenever the JavaScript stack becomes empty](https://youtu.be/cCOL7MC4Pl0?t=1475):
+
```

Comment:
one trick is if you leave `https://youtu.be/cCOL7MC4Pl0?t=1475` as a separate line, BFE.dev would render the youtube video widget.

Reply:
should I open a new PR for this?